### ISSUE TITLE
Introduction of `Gate` abstract class and implementation of `CorrectGate`, with unified interface (`ktg/gates.py`)

### ISSUE THREAD

**Issue by yukiharada1228** [yukiharada1228/KnowledgeTransferGraph#1]
### タイトル
Gate 抽象クラス導入と CorrectGate 実装、インターフェース統一（`ktg/gates.py`）

### 背景
- 現状の `ktg/gates.py` は簡易ゲートのみで、抽象クラスがなくインターフェースが統一されていない。
- `CorrectGate` が未実装のため、教師の正誤に基づくゲーティングが使えない。
- 将来の拡張や一貫性のため、ゲートのインターフェースを単純化・統一したい。

### 目的
- 抽象基底 `BaseGate` を導入して全ゲートのインターフェースを統一する。
- すべてのゲートの `forward` シグネチャを「loss 以外は追加情報（kwargs）」に統一する。
- `CorrectGate` を実装する（追加情報が無い場合はエラーを送出）。
- `LinearGate` は `epoch` と `max_epoch` を kwargs で受け取り、未指定時はエラーを送出。

### 対象範囲
- `ktg/gates.py`
- 必要に応じて呼び出し側（ゲートの初期化・呼び出し）を最小限調整

### 仕様（統一インターフェース）
- すべてのゲートは次のインターフェースに準拠する:
```python
class BaseGate(nn.Module, ABC):
    def forward(self, loss: torch.Tensor, **kwargs) -> torch.Tensor:
        ...
```
- 追加情報はすべて kwargs で渡す:
  - LinearGate: `epoch`, `max_epoch`
  - CorrectGate: `teacher_logits` (N, C), `label_id` (N,)

### 実装要件
- BaseGate を新規追加。
- ThroughGate/CutoffGate/LinearGate を BaseGate 継承・`forward(loss, **kwargs)` に変更。
- CorrectGate を新規実装（必須情報未指定時は `ValueError`）。
- max_epoch はコンストラクタ引数ではなく kwargs として扱う。
- 破壊的変更の周知（旧 `forward(loss, epoch)` → 新 `forward(loss, epoch=..., max_epoch=...)`）。

### 参考実装（抜粋）
```python
from abc import ABC, abstractmethod
import torch, torch.nn as nn

class BaseGate(nn.Module, ABC):
    def __init__(self): super().__init__()
    @abstractmethod
    def forward(self, loss: torch.Tensor, **kwargs) -> torch.Tensor: ...

class ThroughGate(BaseGate):
    def forward(self, loss, **kwargs): return loss

class CutoffGate(BaseGate):
    def forward(self, loss, **kwargs): return torch.zeros_like(loss, requires_grad=True)

class LinearGate(BaseGate):
    def forward(self, loss, **kwargs):
        epoch = kwargs.get("epoch"); max_epoch = kwargs.get("max_epoch")
        if epoch is None or max_epoch is None: raise ValueError("LinearGate requires 'epoch' and 'max_epoch'.")
        if max_epoch <= 0: raise ValueError("LinearGate requires max_epoch > 0.")
        w = torch.as_tensor(epoch, dtype=loss.dtype, device=loss.device) / \
            torch.as_tensor(max_epoch, dtype=loss.dtype, device=loss.device)
        return loss * w

class CorrectGate(BaseGate):
    def forward(self, loss, **kwargs):
        teacher_logits = kwargs.get("teacher_logits"); label_id = kwargs.get("label_id")
        if teacher_logits is None or label_id is None: raise ValueError("CorrectGate requires 'teacher_logits' and 'label_id'.")
        if loss.dim() == 0: raise ValueError("CorrectGate requires per-sample loss (dim > 0).")
        if teacher_logits.dim() != 2: raise ValueError("teacher_logits must be (N, C).")
        label_id = label_id.view(-1).to(teacher_logits.device)
        if teacher_logits.size(0) != label_id.size(0) or loss.size(0) != label_id.size(0):
            raise ValueError("Batch size mismatch among loss/teacher_logits/label_id.")
        with torch.no_grad():
            mask = (teacher_logits.argmax(dim=1) == label_id)
        mask = mask.to(loss.device, loss.dtype).view((mask.size(0),) + (1,) * (loss.dim() - 1))
        return loss * mask
```

### 使い方例
- 線形スケジューリング: `gate(loss, epoch=cur_epoch, max_epoch=total_epochs)`
- Correct: `gate(per_sample_loss, teacher_logits=teacher_logits, label_id=labels)`

### タスクリスト
- [x] `BaseGate` の追加とインターフェース定義
- [x] `ThroughGate`/`CutoffGate`/`LinearGate` の継承化・シグネチャ変更
- [x] `CorrectGate` の新規実装
- [x] 既存呼び出し側の引数更新（`epoch`, `max_epoch` を kwargs で渡す）
- [x] 最低限のテスト（下記）

### 受け入れ条件
- すべてのゲートが `forward(loss, **kwargs)` で動作する
- `LinearGate` に `epoch`/`max_epoch` がない場合・`CorrectGate` に必須情報がない場合に `ValueError` が発生
- `CorrectGate` が教師正解サンプルのみ損失を通すことを確認（per-sample でのマスク適用）

### テスト観点
- ThroughGate/CutoffGate/LinearGate/CorrectGate の基本動作
- `LinearGate` の重みが `epoch/max_epoch` に一致
- `CorrectGate` で教師正解のみ通る（N 次元 loss に対するブロードキャストも検証）
- 必須情報欠如時の例外送出

### 影響範囲 / 互換性
- 破壊的変更: 旧 `forward(loss, epoch)` → 新 `forward(loss, **kwargs)`。`max_epoch` はコンストラクタから kwargs に移動。
- 呼び出し側の引数受け渡し修正が必要。

- **この Issue は `ktg/gates.py` が未整備（抽象化未導入・CorrectGate 未実装）である現状を解消するためのものです。**

- 変更点の要約
  - `BaseGate` を導入し、インターフェースを `forward(loss, **kwargs)` に統一。
  - `CorrectGate` を追加し、教師正解のみ通すゲーティングを実装。
  - `epoch`/`max_epoch` などはすべて追加情報（kwargs）として扱う。
